### PR TITLE
renderer, Nix/HM module: add fade out animation and property to disable

### LINKED ISSUE #60: [REQ] Fade out after unlock
It would be nice to have the option to fade out quickly after successfully unlocking

### PR DESCRIPTION
**Pull request by Sasikuttan2163**
Fades out the lockscreen on successful authentication. Closes #60.

https://github.com/hyprwm/hyprlock/assets/84299080/5e74e3fc-57b8-4ea0-9894-07e80e72bff0

**Note about Nix module: **
I can't test it on Nix as I don't use it. If it doesn't work on Nix please tell me to remove that part, I'd be glad to do so.

### CODE REVIEW

File: src/core/hyprlock.cpp [hunk 1]
```diff
@@ -698,9 +701,14 @@ static const ext_session_lock_v1_listener sessionLockListener = {
 // end session_lock
 
 void CHyprlock::onPasswordCheckTimer() {
+    static auto* const PNOFADEOUT  = (Hyprlang::INT* const*)g_pConfigManager->getValuePtr("general:no_fade_out");
     // check result
     if (m_sPasswordState.result->success) {
-        unlockSession();
+        if (**PNOFADEOUT) {
```
Comment (vaxerski) [inline src/core/hyprlock.cpp:707]: no {} around short ifs

File: src/renderer/Renderer.cpp [hunk 2]
```diff
@@ -199,6 +201,10 @@ CRenderer::SRenderFeedback CRenderer::renderLock(const CSessionLockSurface& surf
         if (**PNOFADEIN)
             bga = 1.0;
 
+        if (g_pHyprlock->m_bFadeStarted && !**PNOFADEOUT) {
+            bga = std::clamp(std::chrono::duration_cast<std::chrono::microseconds>(g_pHyprlock->m_tFadeEnds - std::chrono::system_clock::now()).count() / 1000000.0 - 0.02, 0.0, 1.0);
```
Comment (vaxerski) [inline src/renderer/Renderer.cpp:205]: too long, maybe 0.5s?

File: src/core/hyprlock.cpp [hunk 3]
```diff
@@ -698,9 +701,14 @@ static const ext_session_lock_v1_listener sessionLockListener = {
 // end session_lock
 
 void CHyprlock::onPasswordCheckTimer() {
+    static auto* const PNOFADEOUT  = (Hyprlang::INT* const*)g_pConfigManager->getValuePtr("general:no_fade_out");
```
Comment (vaxerski) [inline src/core/hyprlock.cpp:704]: yes, but if XDG_CURRENT_DESKTOP != Hyprland we should force it to off.

Comment (Sasikuttan2163) [reply src/core/hyprlock.cpp:704]: I'll make the changes, thanks!

File: src/renderer/Renderer.cpp (hunk 2 above)
Comment (Sasikuttan2163) [reply src/renderer/Renderer.cpp:205]: I tried 0.5s but there was a small jump at the end, 0.6s feels perfect though

Comment (Sasikuttan2163) [reply src/renderer/Renderer.cpp]: Here's how it looks like:
https://github.com/hyprwm/hyprlock/assets/84299080/a706a300-ff1d-4c27-8fbd-32a8b5caf1e7

**Review by fufexan — APPROVED** [review]
Nix part LGTM.

File: src/renderer/Renderer.cpp (hunk 2 above)
Comment (vaxerski) [reply src/renderer/Renderer.cpp]: yea better

File: src/renderer/Renderer.cpp [hunk 4]
```diff
@@ -210,7 +216,7 @@ CRenderer::SRenderFeedback CRenderer::renderLock(const CSessionLockSurface& surf
 
     Debug::log(TRACE, "frame {}", frames);
 
-    feedback.needsFrame = feedback.needsFrame || !asyncResourceGatherer->ready || bga < 1.0;
+    feedback.needsFrame = feedback.needsFrame || !asyncResourceGatherer->ready || bga != 1.0;
```
Comment (vaxerski) [inline src/renderer/Renderer.cpp:219]: what's the reason behind this

Comment (Sasikuttan2163) [reply src/renderer/Renderer.cpp:219]: I was making a few changes while debugging, forgot to change that, I'll switch it back. Thanks for pointing it out!

**Comment by Sasikuttan2163** [comment]
I updated the fade to 500 ms and reverted the needsFrame condition, I think all should be good now

**Comment by Sasikuttan2163** [comment]
The last commit fixes the jump at the end. Final result: 

https://github.com/hyprwm/hyprlock/assets/84299080/1c52e641-8222-4a15-9ceb-3eb828447469

**Review by vaxerski — APPROVED** [review]

**Pull request merged**